go/multipooler/manager/rpc_consensus_test.go

```diff
@@ -0,0 +1,780 @@
+
+	"github.com/multigres/multigres/go/clustermetadata/topo/memorytopo"
+	"github.com/multigres/multigres/go/cmd/pgctld/testutil"
+	"github.com/multigres/multigres/go/servenv"
+
+	clustermetadatapb "github.com/multigres/multigres/go/pb/clustermetadata"
+	consensusdatapb "github.com/multigres/multigres/go/pb/consensusdata"
+	pgctldpb "github.com/multigres/multigres/go/pb/pgctldservice"
+)
+
+// TestBeginTerm_AlreadyVotedInOlderTerm tests the scenario where:
+// 1. Node has voted for candidate A in term 5
+// 2. Candidate B requests vote in term 10 (newer term)
+// Expected: Vote should be accepted because the newer term should reset the vote
+// Bug: Currently the code checks votedFor before updating the term, which would reject the vote
+func TestBeginTerm_AlreadyVotedInOlderTerm(t *testing.T) {
+	ctx := context.Background()
+	pm, mock, tmpDir := setupManagerWithMockDB(t)
+
+	// Step 1: Initialize term to 5 and vote for "candidate-A"
+	initialTerm := &pgctldpb.ConsensusTerm{
+		CurrentTerm: 5,
+		VotedFor: &clustermetadatapb.ID{
+			Cell: "zone1",
+			Name: "candidate-A",
+		},
+	}
+	err := SetTerm(tmpDir, initialTerm)
+	require.NoError(t, err)
+
+	// Reload the term in the manager
+	pm.mu.Lock()
+	pm.consensusTerm = initialTerm
+	pm.mu.Unlock()
+
+	// Step 2: Candidate B requests vote in term 10 (newer term)
+	// Mock expectations for Ping call (health check)
+	mock.ExpectPing()
+
+	// Mock expectations for replication status check
+	// Return recent timestamp to indicate we're caught up
+	recentTime := time.Now().Add(-5 * time.Second)
+	mock.ExpectQuery("SELECT last_msg_receipt_time FROM pg_stat_wal_receiver").
+		WillReturnRows(sqlmock.NewRows([]string{"last_msg_receipt_time"}).AddRow(recentTime))
+
+	req := &consensusdatapb.BeginTermRequest{
+		Term:        10, // Newer term
+		CandidateId: "candidate-B",
+		ShardId:     "shard-1",
+	}
+
+	resp, err := pm.BeginTerm(ctx, req)
+
+	// Should succeed
+	require.NoError(t, err)
+	require.NotNil(t, resp)
+
+	// BUG EXPOSED: The vote should be accepted because term 10 is newer than term 5
+	// The code should:
+	// 1. See that req.Term (10) > currentTerm (5)
+	// 2. Update the term to 10 and reset votedFor to nil
+	// 3. Grant the vote to candidate-B
+	//
+	// However, with the current bug, the code checks votedFor != nil && votedFor != "candidate-B"
+	// BEFORE updating the term, so it incorrectly rejects the vote.
+	assert.True(t, resp.Accepted, "Vote should be accepted when request term is newer than current term, even if already voted in older term")
+	assert.Equal(t, int64(10), resp.Term)
+
+	// Verify the vote was persisted
+	loadedTerm, err := GetTerm(tmpDir)
+	require.NoError(t, err)
+	assert.Equal(t, int64(10), loadedTerm.CurrentTerm)
+	assert.Equal(t, "candidate-B", loadedTerm.VotedFor.GetName())
+
+	// Verify all expectations were met
+	assert.NoError(t, mock.ExpectationsWereMet())
+}
+
+// TestBeginTerm_AlreadyVotedInSameTerm tests the scenario where:
```

Comment:
Idea: for these suite, we can rely more in unit tests of the core consensus logic around term validation. I was taking that approach in here: https://github.com/multigres/multigres/blob/main/go/multipooler/manager/manager_test.go#L288 

I think for this kind assertion a matrix style makes it easier to parse all the tests